File: pet-store-repository/src/main/java/com/chewy/pet/repository/DynamoRepositoryInitializer.java
package com.chewy.pet.repository;

import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.DynamoDbClientBuilder;

import java.net.URI;

public class DynamoRepositoryInitializer {

    DynamoConfig config;

    public DynamoRepositoryInitializer(DynamoConfig config) {
        this.config = config;
    }

    public DynamoDbEnhancedClient init() {
        DynamoDbClient client = createClient(config);
        return createEnhancedClient(client);
    }

    public DynamoDbClient createClient(DynamoConfig config) {
        DynamoDbClientBuilder builder = DynamoDbClient.builder();
        if (config.getServiceUrl() != null) {
            builder.endpointOverride(URI.create(config.getServiceUrl()));
            builder.region(Region.of(config.getRegion()));
        }
        return builder.build();
    }

    public DynamoDbEnhancedClient createEnhancedClient(DynamoDbClient dynamoClient) {
        return DynamoDbEnhancedClient.builder()
                .dynamoDbClient(dynamoClient)
                .build();
    }
}

File: pet-store-dynamo-repository/src/main/kotlin/com/johnkruss/osn/dynamo/DynamoRepositoryInitializer.kt
package com.johnkruss.osn.dynamo

import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.dynamodb.DynamoDbClient
import java.net.URI

class DynamoRepositoryInitializer(var config: DynamoConfig) {
    fun init(): DynamoDbEnhancedClient {
        val client = createClient(config)
        return createEnhancedClient(client)
    }

    fun createClient(config: DynamoConfig): DynamoDbClient {
        val builder = DynamoDbClient.builder()
        if (config.serviceUrl != null) {
            builder.endpointOverride(URI.create(config.serviceUrl))
            builder.region(Region.of(config.region))
        }
        return builder.build()
    }

    fun createEnhancedClient(dynamoClient: DynamoDbClient): DynamoDbEnhancedClient {
        return DynamoDbEnhancedClient.builder()
            .dynamoDbClient(dynamoClient)
            .build()
    }
}
